### PR TITLE
Validate datetime

### PR DESCRIPTION
**Pull request by lupasarin**
## _Read then delete this section_

_- Make sure to use a concise title for the pull-request._

_- Use #patch, #minor or #major in the pull-request title to bump the corresponding version. Otherwise, the patch version
will be bumped. [More details](https://github.com/marketplace/actions/github-tag-bump)_

# TL;DR
_Please replace this text with a description of what this PR accomplishes._

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,6 +24,8 @@ Flyte Admin has a few useful make targets for linting and testing. Please use th
 minor bugs and linting errors.
 
 ```
+  # Please make sure you have goimports installed, i.e.
+  $ go install golang.org/x/tools/cmd/goimports@latest
```
Comment (katrogan) [inline README.md:28]: you can actually run `make install` which should download all the go tooling you need to run goimports

**Comment by katrogan** [comment]
thanks @lupasarin for making this change! can you also update [validateLiteralMap](https://github.com/flyteorg/flyteadmin/blob/269b482c45b07352f9c1c78a2b85e14bd85d4266/pkg/manager/impl/validation/validation.go#L193) which is used for launch plan and create execution validation with a similar check?

File: pkg/manager/impl/validation/validation_test.go [hunk 2]
```diff
@@ -334,3 +358,42 @@ func TestValidateOutputData(t *testing.T) {
...
+					Value: &core.Scalar_Primitive{
+						Primitive: &core.Primitive{
+							Value: &core.Primitive_Datetime{Datetime: timestamppb.New(timestamp)},
+						},
+					},
+				},
+			},
+		}))
+	})
+	// cannot generate invalid timestamp with wrong format or e.g. too big month or year.
```
Comment (katrogan) [inline pkg/manager/impl/validation/validation_test.go:397]: one option here is to generate the invalid timestamp proto from your java application, dump this to json and save that json blob as a test file in this repo. Then, to implement this test case you can parse the proto from json using [jsonpb Unmarshal](https://pkg.go.dev/github.com/golang/protobuf/jsonpb) and validate that

File: pkg/manager/impl/validation/validation.go [hunk 3]
```diff
@@ -268,3 +277,38 @@ func ValidateOutputData(outputData *core.LiteralMap, maxSizeInBytes int64) error
...
+	err := timestamp.CheckValid()
+	if err != nil {
+		return errors.NewFlyteAdminErrorf(codes.InvalidArgument, err.Error())
+	}
+	return nil
+}
+
+func isDateTime(input *core.Literal) bool {
+	literalValue := input.Value
+	if reflect.ValueOf(literalValue).String() != "<*core.Literal_Scalar Value>" {
```
Comment (katrogan) [inline pkg/manager/impl/validation/validation.go:297]: instead of using reflection can you just check if the oneof value is nil, something like

`return input.GetScalar() != nil && input.GetScalar().GetPrimitive() != nil && inpuGetScalar().GetPrimitive().GetDatetime() != nil`

File: pkg/manager/impl/validation/validation_test.go (hunk 2 above)
Comment (lupasarin) [reply pkg/manager/impl/validation/validation_test.go]: Thanks for the suggestion! We finally found out we could do this using timestamppb type directly, constructed using Seconds+Nanos.

File: pkg/manager/impl/validation/validation.go (hunk 3 above)
Comment (lupasarin) [reply pkg/manager/impl/validation/validation.go]: Done, thanks for the hint!

File: pkg/manager/impl/validation/validation_test.go (hunk 2 above)
Comment (katrogan) [reply pkg/manager/impl/validation/validation_test.go]: even better, awesome!

File: pkg/manager/impl/validation/validation.go [hunk 4]
```diff
@@ -268,3 +276,21 @@ func ValidateOutputData(outputData *core.LiteralMap, maxSizeInBytes int64) error
 	}
 	return errors.NewFlyteAdminErrorf(codes.ResourceExhausted, "Output data size exceeds platform configured threshold (%+v > %v)", outputSizeInBytes, maxSizeInBytes)
 }
+
+func ValidateDatetime(literal *core.Literal) error {
+	if literal == nil {
+		return nil
+	}
+
+	timestamp := literal.Value.(*core.Literal_Scalar).Scalar.Value.(*core.Scalar_Primitive).Primitive.GetValue().(*core.Primitive_Datetime).Datetime
```
Comment (katrogan) [inline pkg/manager/impl/validation/validation.go:285]:
```suggestion
	timestamp := literal.GetScalar().GetPrimitive().GetDatetime()
```

Comment (katrogan) [reply pkg/manager/impl/validation/validation.go:285]: (also because of the DCO requirement please don't actually accept this suggestion from the Github UI because it won't be signed off 😅 )

**Review by katrogan — DISMISSED** [review]
Thank you for updating all the function comments and refactoring tests into specific, named cases! Just one more suggestion and should be good to go 🙏

File: pkg/manager/impl/validation/validation.go (hunk 4 above)
Comment (lupasarin) [reply pkg/manager/impl/validation/validation.go]: Thanks a lot for the suggestion! Now it's done :)

File: pkg/manager/impl/validation/validation.go [hunk 5]
```diff
@@ -268,3 +276,21 @@ func ValidateOutputData(outputData *core.LiteralMap, maxSizeInBytes int64) error
 	}
 	return errors.NewFlyteAdminErrorf(codes.ResourceExhausted, "Output data size exceeds platform configured threshold (%+v > %v)", outputSizeInBytes, maxSizeInBytes)
 }
+
+func ValidateDatetime(literal *core.Literal) error {
+	if literal == nil {
+		return nil
+	}
```
Comment (ckiosidis) [inline pkg/manager/impl/validation/validation.go:283]: does this nil check make sense?

the fixedInput should not be nil as far as i understand.
This default behaviour will not protect us from future refactorings

Comment (lupasarin) [reply pkg/manager/impl/validation/validation.go:283]: Do you mean that we should fail if literal == nil? Or that it will never happen?

Comment (ckiosidis) [reply pkg/manager/impl/validation/validation.go:283]: I think so and if necessary do refactorings on the places of usage to accomodate for the nil usecase

Comment (lupasarin) [reply pkg/manager/impl/validation/validation.go:283]: As mentioned via private message, I changed the check to throw an error instead. Just to be on the safest side.

File: pkg/manager/impl/validation/validation_test.go [hunk 6]
```diff
@@ -334,3 +359,88 @@ func TestValidateOutputData(t *testing.T) {
...
+						Primitive: &core.Primitive{
+							Value: &core.Primitive_Datetime{Datetime: &timestamp},
+						},
+					},
+				},
+			},
+		}))
+	})
+	t.Run("datetime with value above max", func(t *testing.T) {
+		timestamp := timestamppb.Timestamp{Seconds: 253402300800, Nanos: 0} // = 0000-12-31T23:59:59.999999999Z
```
Comment (ckiosidis) [inline pkg/manager/impl/validation/validation_test.go:429]: let's also add a test for the java minimum (which should not be supported)

```
MIN
The minimum supported Instant, '-1000000000-01-01T00:00Z'.

Comment (lupasarin) [reply pkg/manager/impl/validation/validation_test.go:447]: Done!

**Review by katrogan — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
